shared/warning.ts

```diff
@@ -0,0 +1,124 @@
+import { enumFromStringValue } from './utils';
+
+export enum WarningType {
+  Unknown = 'Unknown',
+  WriteFilePermission = 'WriteFilePermission',
+}
+
+export enum WriteFilePermissionWarningKinds {
+  Logger = 'Logger',
+  ConfigFile = 'ConfigFile',
+  WalletFile = 'WalletFile',
+}
+
+interface WarningTypeOptions {
+  [WarningType.Unknown]: WarningOptions<Record<string, any>>;
+  [WarningType.WriteFilePermission]: WarningOptions<{
+    kind: WriteFilePermissionWarningKinds;
+    filePath: string;
+  }>;
+}
+
+export type WarningPayload<
+  T extends WarningType
+> = WarningTypeOptions[T]['payload'];
+
+type WarningOptions<P> = {
+  message: string;
+  stack?: string;
+  payload: P;
+};
+
+export type WarningObject<T extends keyof WarningTypeOptions> = {
+  type: T;
+  message: string;
+  stack?: string;
+  payload: WarningPayload<T>;
+};
+
+export type AnyWarningObject = WarningObject<WarningType>;
```

Comment:
```suggestion
export type WarningObjectType = WarningObject<WarningType>;
```

Reply:
I think `AnyWarningObject` is a more telling name.
It is a union type and literally represents any of the possible warning objects ;)

However, I checked it again and found that it would be better to refactor it a bit (now it is an unclear union of all possible types and payloads). Will push changes that will make `AnyWarningObject` a union of possible WarningObjects, aka `WarningObject<Unknown> | WarningObjecT<WriteFileError>`.